artichoke-backend/src/extn/core/kernel/integer.rs

```diff
@@ -0,0 +1,177 @@
+use regex::Regex;
+use std::collections::HashMap;
+use std::str::FromStr;
+
+use crate::convert::{Convert, TryConvert};
+use crate::types::Ruby;
+use crate::value::Value;
+use crate::Artichoke;
+
+#[derive(Debug, Clone, Hash, PartialEq, Eq)]
+pub enum Error {
+    BaseSpecifiedForNonString(bool),
+    ContainsNullByte(bool),
+    InvalidValue(String, bool),
+    InvalidRadix(String, bool),
+    NoImplicitConversionToString(String, bool),
+}
+
+#[derive(Debug)]
+pub struct Args {
+    pub arg: String,
+    pub radix: Option<i64>,
+    pub raise_exception: bool,
+}
+
+impl Args {
+    pub unsafe fn extract(interp: &Artichoke) -> Result<Self, Error> {
+        let mrb = interp.0.borrow().mrb;
+
+        // TODO: when `base` is not passed, but `exception: true` is,
+        // exception argument goes into arg, which is not right,
+        // we might want to get the argument manually.
+        let (arg, base, exception) = mrb_get_args!(mrb, required = 1, optional = 2);
+
+        let raise_exception: Option<HashMap<String, bool>> = match exception {
+            Some(exception) => match interp.try_convert(Value::new(interp, exception)) {
+                Ok(exception) => exception,
+                Err(_) => None,
+            },
+            _ => None,
+        };
+        let raise_exception: bool = if let Some(raise_exception) = raise_exception {
+            *raise_exception.get("exception").unwrap_or_else(|| &true)
+        } else {
+            true
+        };
+        let radix: Option<i64> = match base {
+            Some(base) => {
+                if let Ok(base) = interp.try_convert(Value::new(interp, base)) {
+                    base
+                } else {
+                    None
+                }
+            }
+            _ => None,
+        };
+
+        let value = Value::new(interp, arg);
+        if value.ruby_type() == Ruby::Fixnum && radix.is_some() {
+            return Err(Error::BaseSpecifiedForNonString(raise_exception));
+        }
+
+        if let Ok(arg) = interp.try_convert(Value::new(interp, arg)) {
+            Ok(Self {
+                arg,
+                radix,
+                raise_exception,
+            })
+        } else {
+            Err(Error::NoImplicitConversionToString(
+                value.ruby_type().class_name().into(),
```

Comment:
this is not the API we want.

We actually want something like:

https://github.com/artichoke/artichoke/blob/549b8b1df775427d8f8924c9d0ae072a4ca34378/artichoke-backend/src/value.rs#L105-L123